clients/shared/merge.go

```diff
@@ -0,0 +1,164 @@
+	if tableData.ShouldSkipUpdate() {
+		return nil
+	}
+
+	fqName := tableData.ToFqName(dwh.Label(), true, cfg.SharedDestinationConfig.UppercaseEscapedNames, "")
+	tableConfig, err := dwh.GetTableConfig(tableData)
+	if err != nil {
+		return err
+	}
+
+	srcKeysMissing, targetKeysMissing := columns.Diff(tableData.ReadOnlyInMemoryCols(), tableConfig.Columns(),
+		tableData.TopicConfig.SoftDelete, tableData.TopicConfig.IncludeArtieUpdatedAt,
+		tableData.TopicConfig.IncludeDatabaseUpdatedAt, tableData.Mode())
+
+	createAlterTableArgs := ddl.AlterTableArgs{
+		Dwh:               dwh,
+		Tc:                tableConfig,
+		FqTableName:       fqName,
+		CreateTable:       tableConfig.CreateTable(),
+		ColumnOp:          constants.Add,
+		CdcTime:           tableData.LatestCDCTs,
+		UppercaseEscNames: &cfg.SharedDestinationConfig.UppercaseEscapedNames,
+	}
+
+	// Columns that are missing in DWH, but exist in our CDC stream.
+	err = ddl.AlterTable(createAlterTableArgs, targetKeysMissing...)
+	if err != nil {
+		slog.Warn("Failed to apply alter table", slog.Any("err", err))
+		return err
+	}
+
+	// Keys that exist in DWH, but not in our CDC stream.
+	deleteAlterTableArgs := ddl.AlterTableArgs{
+		Dwh:                    dwh,
+		Tc:                     tableConfig,
+		FqTableName:            fqName,
+		CreateTable:            false,
+		ColumnOp:               constants.Delete,
+		ContainOtherOperations: tableData.ContainOtherOperations(),
+		CdcTime:                tableData.LatestCDCTs,
+		UppercaseEscNames:      &cfg.SharedDestinationConfig.UppercaseEscapedNames,
+	}
+
+	if err = ddl.AlterTable(deleteAlterTableArgs, srcKeysMissing...); err != nil {
+		slog.Warn("Failed to apply alter table", slog.Any("err", err))
+		return err
+	}
+
+	tableConfig.AuditColumnsToDelete(srcKeysMissing)
+	tableData.MergeColumnsFromDestination(tableConfig.Columns().GetColumns()...)
+	temporaryTableName := fmt.Sprintf("%s_%s", tableData.ToFqName(dwh.Label(), false, cfg.SharedDestinationConfig.UppercaseEscapedNames, ""), tableData.TempTableSuffix())
+	if err = dwh.PrepareTemporaryTable(tableData, tableConfig, temporaryTableName, types.AdditionalSettings{}); err != nil {
+		return fmt.Errorf("failed to prepare temporary table: %w", err)
+	}
+
+	defer func() {
+		if dropErr := ddl.DropTemporaryTable(dwh, temporaryTableName, false); dropErr != nil {
+			slog.Warn("Failed to drop temporary table", slog.Any("err", dropErr), slog.String("tableName", temporaryTableName))
+		}
+	}()
+
+	// Now iterate over all the in-memory cols and see which one requires backfill.
+	for _, col := range tableData.ReadOnlyInMemoryCols().GetColumns() {
+		if col.ShouldSkip() {
+			continue
+		}
+
+		var attempts int
+		for {
+			err = BackfillColumn(cfg, dwh, col, fqName)
+			if err == nil {
+				tableConfig.Columns().UpsertColumn(col.RawName(), columns.UpsertColumnArg{
+					Backfilled: ptr.ToBool(true),
+				})
+				break
+			}
+
+			if opts.RetryColBackfill && dwh.IsRetryableError(err) {
```

Comment:
Do we not want some max amount of attempts?

Reply:
Doesn't that defeat the purpose of a retryable error?

Reply:
Not at all, it's fine to retry a retryable error several times, maybe 10s or even 100s of times, but if we've retried 1000s of times and it's not completing we should probably fail rather than retrying forever. Even if they are super unlikely, while true loops are the worst to debug.

Reply:
Fair, I can leave a TODO and set the upper bound of attempts to be 1k